stacks-common/src/util/lru_cache.rs

```diff
@@ -0,0 +1,596 @@
+            self.move_to_head(index)?;
+            Ok(None)
+        } else {
+            let mut evicted = None;
+            // This is a new key
+            let index = if self.cache.len() == self.capacity {
+                // We've reached capacity. Evict the least-recently used value
+                // and reuse its node
+                let index = self.evict_lru()?;
+                let tail_node = self.order.get_mut(index).ok_or(LruCacheCorrupted)?;
+
+                // Replace the key with the new key, saving the old key
+                let replaced_key = std::mem::replace(&mut tail_node.key, key.clone());
+
+                // Save the evicted key-value pair, if it was dirty
+                if tail_node.dirty {
+                    evicted = Some((replaced_key, tail_node.value));
+                };
+
+                // Update the evicted node with the new key-value pair
+                tail_node.value = value;
+                tail_node.dirty = dirty;
+
+                // Insert the new key-value pair into the cache
+                self.cache.insert(key, index);
+
+                index
+            } else {
+                // Create a new node, add it to the cache
+                let index = self.order.len();
+                let node = Node {
+                    key: key.clone(),
+                    value,
+                    dirty,
+                    next: self.capacity,
+                    prev: self.capacity,
+                };
+                self.order.push(node);
+                self.cache.insert(key, index);
+                index
+            };
+
+            // Put the new or reused node at the head of the LRU list
+            self.attach_as_head(index)?;
+
+            Ok(evicted)
+        }
+    }
+
+    /// Flush all dirty values in the cache, calling the given function, `f`,
+    /// for each dirty value.
+    /// Outer result is an error iff the cache is corrupted and should be discarded.
+    /// Inner result is an error iff the function, `f`, returns an error.
+    pub fn flush<E>(
+        &mut self,
+        mut f: impl FnMut(&K, V) -> Result<(), E>,
+    ) -> Result<Result<(), E>, LruCacheCorrupted> {
+        let mut current = self.head;
+
+        // Keep track of visited nodes to detect cycles
+        let mut visited = HashSet::new();
+
+        while current != self.capacity {
+            // Detect cycles
+            if !visited.insert(current) {
+                return Err(LruCacheCorrupted);
+            }
+
+            let node = self.order.get_mut(current).ok_or(LruCacheCorrupted)?;
+            let next = node.next;
+            if node.dirty {
+                let value = node.value;
+
+                // Call the flush function
+                match f(&node.key, value) {
+                    Ok(()) => node.dirty = false,
+                    Err(e) => return Ok(Err(e)),
+                }
+                node.dirty = false;
+            }
+            current = next;
```

Comment:
This looks okay as it is right now, but would it be easier to just iterate over `self.order` directly, i.e.:

```rust
for dirty_node in self.order.iter_mut().filter(|n| n.dirty) {
    f(&node.key, value)?;
    node.dirty = false;
}
```

It doesn't seem like it matters to the caller that the flush be executed by LRU order.

Reply:
Good point. We still need to handle the outer `Result` though. Do you like the version with the `match`:

```rust
        for node in self.order.iter_mut().filter(|n| n.dirty) {
            match f(&node.key, node.value) {
                Ok(()) => node.dirty = false,
                Err(e) => return Ok(Err(e)),
            }
        }
        Ok(Ok(()))
```

or this more functional version:

```rust
        let result = self
            .order
            .iter_mut()
            .filter(|node| node.dirty)
            .map(|node| {
                f(&node.key, node.value).map(|_| {
                    node.dirty = false;
                })
            })
            .collect::<Result<(), E>>();
        Ok(result)
```

I think the first is easier to follow.

Reply:
I went with the `match` version in 270437c4ec48ff7becb2c5a34dc2d7a869770324.

Reply:
Oh, I didn't notice the `Result<Result<` here... Now, I'm concerned about that! I think `Result<Result<` should be avoided because it introduces the possibility that a check like this is written:

```rust
if let Err(e) = result {
} else {
   // everything's cool!
}
```

which could be bad (and _has been_ bad before).

I'd prefer something a little more like:

```rust
pub enum <E> FlushError {
   LRUCacheCorrupted,
   FlushError(E)
}
```

And then the return type could be `Result<(), FlushError>`

Reply:
Yeah, you're right. Fixed in 9eba7469224042e6c20a51d557e469c24722737a.